### PR TITLE
Add Observability to Data Model Operations is DynamoDB HLL

### PR DESCRIPTION
**Pull request by 96malhar**
## Description

This PR adds observability to Data Model operations is DynamoDB HLL by creating traces using the SDK observability APIs.
* Traces are only created for public APIs that make a network call.
* I have updated the Async public APIs to await the `Task` instead of returning it. This is done to ensure that the trace is not disposed before the function completes its execution.

## Motivation and Context

This change is required as part of the Observability initiative for the SDK and HLLs
DOTNET-7679

## Testing

* All existing Unit and Integrations tests pass locally
* **Pending a full dry-run**. Since I am targeting a feature branch, I'll perform a full dry run on the feature branch itself after observability updates are merged in for both the data model and document model operations

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] My code follows the code style of this project
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [x] I have read the **README** document
- [ ] I have added tests to cover my changes
- [x] All new and existing tests passed

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

[issues]: https://github.com/aws/aws-sdk-net/issues
[license]: http://aws.amazon.com/apache2.0/
[cla]: http://en.wikipedia.org/wiki/Contributor_License_Agreement

### CODE REVIEW

File: sdk/src/Services/DynamoDBv2/Custom/DataModel/TransactWrite.cs [hunk 1]
```diff
@@ -466,13 +470,15 @@ public partial class MultiTableTransactWrite
         public MultiTableTransactWrite(params TransactWrite[] transactionParts)
         {
             allTransactionParts = new List<TransactWrite>(transactionParts);
+            TracerProvider = allTransactionParts[0].TracerProvider;
```
Comment (normj) [inline sdk/src/Services/DynamoDBv2/Custom/DataModel/TransactWrite.cs:473]: For defensive code purposes can we check if the collection is empty and if it is use the default trace provider from `AWSConfigs.TelemetryProvider`?

**Review by normj — CHANGES_REQUESTED** [review]
@philasmar PR for adding observability for S3 TransferUtility is stripping the `Async` suffix and your PR keeps the `Async` suffix. Can you 2 discuss what should be our pattern be.

File: sdk/src/Services/DynamoDBv2/Custom/DataModel/BatchGet.cs [hunk 2]
```diff
@@ -258,13 +264,15 @@ public partial class MultiTableBatchGet
...
             allBatches = new List<BatchGet>(batches);
+            TracerProvider = allBatches[0].TracerProvider;
         }
 
         internal MultiTableBatchGet(BatchGet first, params BatchGet[] rest)
         {
             allBatches = new List<BatchGet>();
             allBatches.Add(first);
             allBatches.AddRange(rest);
+            TracerProvider = allBatches[0].TracerProvider;
```
[the remark this answers is not in the document]
Comment (96malhar) [reply sdk/src/Services/DynamoDBv2/Custom/DataModel/BatchGet.cs:279]: Done

File: sdk/src/Services/DynamoDBv2/Custom/DataModel/BatchWrite.cs [hunk 3]
```diff
@@ -255,13 +262,15 @@ public partial class MultiTableBatchWrite
         public MultiTableBatchWrite(params BatchWrite[] batches)
         {
             allBatches = new List<BatchWrite>(batches);
+            TracerProvider = allBatches[0].TracerProvider;
```
[the remark this answers is not in the document]
Comment (96malhar) [reply sdk/src/Services/DynamoDBv2/Custom/DataModel/BatchWrite.cs]: Done

File: sdk/src/Services/DynamoDBv2/Custom/DataModel/BatchWrite.cs [hunk 4]
```diff
@@ -255,13 +262,15 @@ public partial class MultiTableBatchWrite
...
         }
 
         internal MultiTableBatchWrite(BatchWrite first, params BatchWrite[] rest)
         {
             allBatches = new List<BatchWrite>();
             allBatches.Add(first);
             allBatches.AddRange(rest);
+            TracerProvider = allBatches[0].TracerProvider;
```
[the remark this answers is not in the document]
Comment (96malhar) [reply sdk/src/Services/DynamoDBv2/Custom/DataModel/BatchWrite.cs:275]: Done

File: sdk/src/Services/DynamoDBv2/Custom/DataModel/TransactGet.cs [hunk 5]
```diff
@@ -248,13 +254,15 @@ public partial class MultiTableTransactGet
         public MultiTableTransactGet(params TransactGet[] transactionParts)
         {
             allTransactionParts = new List<TransactGet>(transactionParts);
+            TracerProvider = allTransactionParts[0].TracerProvider;
```
[the remark this answers is not in the document]
Comment (96malhar) [reply sdk/src/Services/DynamoDBv2/Custom/DataModel/TransactGet.cs]: Done

File: sdk/src/Services/DynamoDBv2/Custom/DataModel/TransactGet.cs [hunk 6]
```diff
@@ -248,13 +254,15 @@ public partial class MultiTableTransactGet
...
         }
 
         internal MultiTableTransactGet(TransactGet first, params TransactGet[] rest)
         {
             allTransactionParts = new List<TransactGet>();
             allTransactionParts.Add(first);
             allTransactionParts.AddRange(rest);
+            TracerProvider = allTransactionParts[0].TracerProvider;
```
[the remark this answers is not in the document]
Comment (96malhar) [reply sdk/src/Services/DynamoDBv2/Custom/DataModel/TransactGet.cs:267]: Done

File: sdk/src/Services/DynamoDBv2/Custom/DataModel/TransactWrite.cs (hunk 1 above)
Comment (96malhar) [reply sdk/src/Services/DynamoDBv2/Custom/DataModel/TransactWrite.cs]: Done

File: sdk/src/Services/DynamoDBv2/Custom/DataModel/TransactWrite.cs [hunk 7]
```diff
@@ -466,13 +470,15 @@ public partial class MultiTableTransactWrite
...
         }
 
         internal MultiTableTransactWrite(TransactWrite first, params TransactWrite[] rest)
         {
             allTransactionParts = new List<TransactWrite>();
             allTransactionParts.Add(first);
             allTransactionParts.AddRange(rest);
+            TracerProvider = allTransactionParts[0].TracerProvider;
```
[the remark this answers is not in the document]
Comment (96malhar) [reply sdk/src/Services/DynamoDBv2/Custom/DataModel/TransactWrite.cs:483]: Done

**Comment by 96malhar** [comment]
> @philasmar PR for adding observability for S3 TransferUtility is stripping the `Async` suffix and your PR keeps the `Async` suffix. Can you 2 discuss what should be our pattern be.

Added custom logic to strip the `Async` suffix

File: sdk/src/Services/DynamoDBv2/Custom/DataModel/BatchGet.cs [hunk 8]
```diff
@@ -258,13 +264,15 @@ public partial class MultiTableBatchGet
         public MultiTableBatchGet(params BatchGet[] batches)
         {
             allBatches = new List<BatchGet>(batches);
+            TracerProvider = allBatches[0].TracerProvider;
```
[the remark this answers is not in the document]
Comment (96malhar) [reply sdk/src/Services/DynamoDBv2/Custom/DataModel/BatchGet.cs]: The CRUD operations (BatchGet, BatchWrite, TransactGet, TransactWrite) in DynamoDB HLL all have internal constructors.
For customers to initialize them, they need to call the `CreateXXX` method on the `DynamoDBContext` class. [Example](https://github.com/aws/aws-sdk-net/blob/ca4005485bca2295d2985250ac25e774f01e2c4d/sdk/src/Services/DynamoDBv2/Custom/DataModel/Context.cs#L219)

Customers will typically use a single instance of the `DynamoDBContext` object per region within their application. 
Thus all CRUD objects will be initialized with the same context and will have the same `TraceProvider` instance

**Review by normj — APPROVED** [review]

**Review by philasmar — APPROVED** [review]

**Pull request merged**